### ISSUE TITLE
[Bug][Routing] "format" treated as part of media "id"

### ISSUE THREAD

**Issue by NeuralClone** [sonata-project/SonataMediaBundle#799]
When interpreting download routes for media, the `format` parameter is currently treated as part of the `id` parameter due to the requirement placed on ids:

``` xml
<route id="sonata_media_download" path="/download/{id}/{format}">
    <default key="_controller">SonataMediaBundle:Media:download</default>
    <default key="format">reference</default>
    <requirement key="id">.*</requirement>
</route>
```

``` xml
<route id="sonata_media_gallery_view" path="/view/{id}">
    <default key="_controller">SonataMediaBundle:Gallery:view</default>
    <requirement key="id">.*</requirement>
</route>
```

This is less of an issue with galleries since an incorrect route will cause a `NotFoundHttpException` to be thrown as expected (although I find this lax requirement rather odd). However, this becomes problematic when attempting to download media when a format other than reference is specified because _valid_ routes will also cause a `NotFoundHttpException` exception to be thrown. Given this undesirable and unexpected behavior, this seems like a bug to me.

The rather lax requirement on the `id` parameter seems a bit strange since media and gallery ids are defined as integers in their respective model mappings. Changing the requirements on `id` to `\d+` in both routes fixes this problem. It's easy enough to override the default MediaBundle routes in the ApplicationSonataMediaBundle that gets generated but that shouldn't be necessary to get those routes to work. This change also shouldn't be a BC break (as far as I can tell) unless people have written code to rely on this behavior.

A specific example of this bug in action:

For a file with a media id of 100 and a format big (default context), the following route should be generated if requesting to download it: `/<some prefix>/download/100/default_big`. Instead of `100` being interpreted as the `id` and `default_big` being interpreted as the `format`, `100/default_big` is treated as the `id`, which is obviously incorrect.

Am I correct in assuming that this is not the desired behavior? If this _is_ the desired behavior, then it seems to render the `format` parameter completely pointless without additional code to interpret the `id` and `format` parameters, and that would seem like a rather roundabout way to deal with the parameters when the Routing component can handle most of the heavy lifting.

**Comment by core23** [comment]
Can you provide a PR for that @NeuralClone ?

**Comment by NeuralClone** [comment]
@core23 Sure, I can work on that. Which branch should I create the PR on?

**Comment by greg0ire** [comment]
> Which branch should I create the PR on? 3.x?

Yes, and if you do, please make sure there is no BC-break.

**Comment by NeuralClone** [comment]
@greg0ire Ok, sounds good!

**Comment by OskarStark** [comment]
Thank you @NeuralClone 👍

**Comment by NeuralClone** [comment]
I'm really sorry about the delay in getting back to this! It's been a busy month.

It looks like the original solution I came up with last year **does** cause a BC break. `id` is a string in both routes, not an integer. So the strict requirement I suggested will break code that correctly treats `id` as a string.

However, there's simpler fix. I'm going to suggest we just remove the requirement on `id` for both of the affected routes. Both [Sonata\MediaBundle\Controller\MediaController::downloadAction()](https://github.com/sonata-project/SonataMediaBundle/blob/91070f85cba5fc262cbf284711f98b4952153d9a/Controller/MediaController.php#L45) and [Sonata\MediaBundle\Controller\GalleryController::viewAction()](https://github.com/sonata-project/SonataMediaBundle/blob/91070f85cba5fc262cbf284711f98b4952153d9a/Controller/GalleryController.php#L35) expect `id` to be a string. The Routing component should take care of both of those automatically without specifying any requirements on `id`.

As I mentioned in the original post, the source of the bug is the `.*` regular expression in the requirement fields. If that's removed the route should still match but the Routing component should only grab what's between the forward slashes instead of capturing everything.

I'm going to double check all of this to make sure everything works as expected and my fix does what I think it does before I open a pull request. In the meantime, does this solution sound acceptable? Or at the very least, a step in the right direction? Unless I'm missing something obvious, of course. :smile:

**Comment by NeuralClone** [comment]
As a quick follow-up to the above, I now I see why I set a stricter requirement on `id`. The entity metadata defaults to `id` being an integer but the functions in the appropriate controllers treat `id` as a string.

[Resources/config/doctrine/Media.orm.xml.skeleton](https://github.com/sonata-project/SonataMediaBundle/blob/91070f85cba5fc262cbf284711f98b4952153d9a/Resources/config/doctrine/Media.orm.xml.skeleton#L16)
[Resources/config/doctrine/Gallery.orm.xml.skeleton](https://github.com/sonata-project/SonataMediaBundle/blob/91070f85cba5fc262cbf284711f98b4952153d9a/Resources/config/doctrine/Gallery.orm.xml.skeleton#L16)

So it looks like treating `id` as a string in the route is still probably the best approach to prevent a BC break. It does create a bit of an inconsistency when it comes to the bundle's defaults but it also makes things more flexible.

**Comment by greg0ire** [comment]
Sounds good to me. Maybe have a look at the git log to see if there is an explanation for the current requirement?

**Comment by NeuralClone** [comment]
@greg0ire Sure, I'll take a closer look at that. I'd hate to make this change only to re-introduce another problem that was fixed a few years ago.

**Comment by NeuralClone** [comment]
Ok, I looked into this some more and apparently the requirement was added on both of the affected routes to fix an issue with PHPCR-ODM: https://github.com/sonata-project/SonataMediaBundle/commit/a88082e1940c5828b06a11caf5b3ad2e582dd3a1.

So I'm not quite sure what the best approach is to fixing this now.

**Comment by greg0ire** [comment]
What about adding a `[^/] ` requirement to `id`? cc @dantleech

**Comment by greg0ire** [comment]
ping @NeuralClone , did I say something stupid, or is there something I need to clarify?

**Comment by NeuralClone** [comment]
@greg0ire Not stupid at all. :) For some reason I didn't get a notification about your first follow-up comment. It wasn't my intention to ignore you. Sorry about that!

I'll give that suggestion a try.

**Comment by greg0ire** [comment]
Great. Awesome! Great. Looking forward to that :)

**Comment by greg0ire** [comment]
Not so sure about that after all. I think "full paths" as referred to in a88082e. So maybe the solution would be to add a constraint on format instead, ensuring it can't be empty, because that's what is happening when the bug happens, right? id eats format when it shouldn't.

Long story short, if I understood correctly, we want `download/some/path/json` to give `some/path` as an id and `json` as a format, so my first suggestion is obviously wrong.

**Comment by NeuralClone** [comment]
Yeah, I believe that's correct.

The `.+` requirement on format (in addition to the existing requirement, `.*`, on id) appears to work. I'm going to do some more tests to make sure things are working as they should but right now it's looking like a good fix. 👍

**Comment by NeuralClone** [comment]
Ok, the `.+` requirement on format definitely seems to do the trick with the `sonata_media_download"` route. Thank you for that suggestion! It's simple and it shouldn't cause a BC break.

I've been trying to reproduce the bug with the `sonata_media_gallery_view` route as stated in my original post and it doesn't seem to be an issue anymore as far as I can tell. I'm not entirely sure why I was having a problem with that route last year because the existing requirement should work since the route hasn't changed in 2 years.

It's possible I'm missing something but it doesn't look like that particular route is affected after all. I may have erroneously thought it was broken.

**Comment by greg0ire** [comment]
Software evolves a lot, it's possible that the Symfony router has changed in two years regarding this. It would be great to have your tests automated by writing a container-aware functional test that would load the router and see if the matching gives the correct values for every parameter in every situation

**Comment by NeuralClone** [comment]
Yeah, that's a very good point. Whatever the case, it's no longer a problem. :)

And yes, I'll definitely add automated tests for that. I'll try to get a pull request up in the next day or so.

**Comment by NeuralClone** [comment]
Unfortunately, after some more testing with better tests, the proposed fix doesn't seem to work after all.

For example:
```php
$testParams2 = array(
    '_controller' => 'SonataMediaBundle:Media:download',
    'format'      => 'default_big',
    'id'          => 'media-id',
    '_route'      => 'sonata_media_download',
);

$this->assertSame($testParams2, $this->router->match('/download/media-id/default_big'));
```

`Router::match()` is returning the following array, which is definitely incorrect for what we want:
```php
array(
    '_controller' => 'SonataMediaBundle:Media:download',
    'format'      => 'reference',
    'id'          => 'media-id/default_big',
    '_route'      => 'sonata_media_download',
)
```

Everything works for the default case of `/download/someid`. If a format is specified, then the id and format are treated as part of the id. For example, if we have a path such as `/download/someid/someformat` then the id will be `someid/someformat` instead of simply `someid`.

(Also, please ignore the above references to commits. I forgot that references to issues automatically show up even without opening a pull request. The tests I'm using now are much better.)

**Comment by greg0ire** [comment]
> If a format is specified

You mean it's possible not to specify a format? This would make thing impossible to solve…

**Comment by NeuralClone** [comment]
Ok, I think I might be confused what's supported. :)

Should [MediaController::downloadAction()](https://github.com/sonata-project/SonataMediaBundle/blob/91070f85cba5fc262cbf284711f98b4952153d9a/Controller/MediaController.php#L45) **always** download files using the reference format? Or should it be possible to specify a format and then be served that version of it? In other words, can the thumbnail versions of images that are created be downloaded?

Let's say I upload an image called map.jpg with an id of 2. The image is then resized in big and small formats. Should the paths `/download/2/small` and `/download/2/big` be valid paths? Or is only `/download/2` supported? Should `/download/2/reference` work?

If the reference format should always be used when downloading, then it's my test that's wrong and will always fail because the feature isn't supported. Is this making any sense?

**Comment by greg0ire** [comment]
It feels like you read this : 

> You mean it's possible to specify a format?

instead of this : 

> You mean it's possible not to specify a format?

I think every url you showed should work, but I don't know this is even possible. What if the id is `whatever/small` instead of `2`? There is no way to tell if the format is part of the id or not.

**Comment by NeuralClone** [comment]
Thanks. I think that answers my questions. :)

I'll revamp my tests a bit as some of my assumptions about how the routes should work were flawed. Assuming that all goes well I'll open a pull request with the changes later tonight.

**Comment by greg0ire** [comment]
> I'll revamp my tests a bit as some of my assumptions about how the routes should work were flawed. 

Good call because right now, with this url pattern, it can't work IMO.

Not sure if moving the `format` parameter to the query string should be considered a BC-break, since at the moment, it is never filled…

**Comment by NeuralClone** [comment]
Yeah, that's exactly what was confusing me. I was operating under the assumption that `format` should be getting filled when it actually isn't. What makes it more confusing is that `format` gets filled when using the view route.

Unless paths are hardcoded in someone's controller or templates, and that's a very bad practice, I'm leaning towards it not being a BC-break to move `format` to the query string. The functionality and signature of the controller should remain the same with that change.

**Comment by NeuralClone** [comment]
I'm really sorry I keep dragging out this discussion (and for the length of this post). I only want what's best for this bundle and this is proving to be a rather frustrating issue to fix. :)

Sadly, after reviewing my original post, thinking about this some more, and playing with various test cases, this issue doesn't appear to be resolved by adding a `.+` requirement to `format`.

Right now the requirement on `id` allows any character to be used, including /. I've written tests to include multiple / characters in paths and they're **always** included as part of the `id` parameter. This is also the case when the `.+` requirement is placed on `format`. `format` is never filled because the `.*` requirement on `id` will always match any character. So the requirement on `format` is effectively ignored.

The way I see it, not allowing / in ids should solve this problem since the / would act as a delimiter. In other words, the first fix @greg0ire suggested above (i.e., add a `[^/]` or similar requirement to `id`) seems to do the trick. This change would still allow any other character to be used in ids.

Not allowing / characters in ids might cause a BC-break and it might reintroduce the problem that was fixed in https://github.com/sonata-project/SonataMediaBundle/commit/a88082e1940c5828b06a11caf5b3ad2e582dd3a1. Admittedly, I'm not familiar enough with PHPCR-ODM to know for certain about that last part. It's also unclear if the intent is to allow ids to use / characters or not due to the inconsistency in the existing route requirements.

**Summary**
Allowing the / character in ids with the current route patterns causes two problems:

1) For the `sonata_media_view` route, the `id` parameter can never contain a /. Everything after the / will **always** be considered part of the `format`. In situations where `id` contains a /, this will result in incorrect paths and exceptions will be thrown.

2) For the `sonata_media_download` route, a similar problem occurs. If the `id` parameter contains a / character, the `format` parameter will **never** be filled. A single / character, even if it isn't part of `id`, will still be treated as part of `id` by the routing matcher. Multiple / characters will result in the same issue.

The end result is a major inconsistency in how `id` is treated.

**Comment by greg0ire** [comment]
Uuuuu… what about moving format to the query string? That would solve everything, right? In `/download/my/phpcr/odm/id?format=big`, there is no possible confusion between id and format, is it?

**Comment by NeuralClone** [comment]
That should work in theory. However, there isn't a way to include query strings in routing configuration. I had looked for that in the past and the feature has been rejected multiple times due to security concerns.

You can manually append query strings and then check the values by manipulating the request. The only way I know of to capture query string values is through the RequestStack service, and that would cause a BC-break.

```php
public function downloadAction($id, RequestStack $requestStack)
{
    $format = $requestStack->getCurrentRequest()->query->get('format');

    if (empty($format)) {
        $format = 'reference';
    }
    // ...
}
```

```php
$this->router->generate('sonata_media_download', array(
    'id'     => 'my/phpcr/odm/id',
    'format' => 'big',
));
```

```twig
{{ path('sonata_media_download', {'id': 'my/phpcr/odm/id', 'format': 'big'}) }}
```

Both of those should produce `/download/my/phpcr/odm/id?format=big`.

**Comment by greg0ire** [comment]
Yes, if you do that you won't see format in the routing configuration.

> That kind of feature has been rejected multiple times due to security concerns.

What are you talking about? Can you give an example? :confused: 

> You can manually append query strings and then check the values by manipulating the request

Yes, and it's completely ok to do so… can you imagine making a search page without query strings? [Symfony docs itself uses them](http://symfony.com/search?q=test)

> The only way I know of to capture query string values is through the RequestStack service, and that would cause a BC-break.

Yeah but a BC-break from sth that doesn't work anyway is maybe not so serious.

**Comment by greg0ire** [comment]
My point is, at the moment `$format` seems to be ignored, so if you remove it from the route, the result will be just the same, right?

**Comment by greg0ire** [comment]
From a dicussion with @lsmith77 on slack : 

>  think we deprecated the PHPCR ODM integration of the SonataMediaBundle
@dbu probably knows more there

**Comment by greg0ire** [comment]
If this is true, then the fix is simple : simply revert to forbidding slashes…

**Comment by NeuralClone** [comment]
> What are you talking about? Can you give an example?

Sorry, I was tired when I posted that and didn't clarify as much as I should have. I just meant that Symfony doesn't support query strings when configuring routes. People have requested that as a feature in the past and it's been repeatedly rejected.

You can obviously use query strings with Symfony. I wasn't trying to say using query strings is bad or anything. They're perfectly safe if you use them properly. I was only trying to say that we'd need to make slightly larger (albeit trivial) changes if we were to use them.

> Yeah but a BC-break from sth that doesn't work anyway is maybe not so serious.
> My point is, at the moment $format seems to be ignored, so if you remove it from the route, the result will be just the same, right?

Yeah, good points. I also don't think this BC-break would be an issue since the feature isn't working properly anyway and it hasn't been for over a year.

> If this is true, then the fix is simple : simply revert to forbidding slashes…

Oh wow. If that's the case, then we've looped back to your first suggestion again. 😄 If we can get a confirmation on that then it shouldn't take me too long to adapt what I've already written.

**Comment by dbu** [comment]
i would not be aware of a decision to deprecate the phpcr-odm integration here. what we atm don't maintain anymore is the https://github.com/symfony-cmf/media-bundle which builds on top of the sonata media bundle.

in phpcr-odm, document ids indeed are paths with forward slashes. the alternative would be to use the uuid of the documents which is a uuid v4. this will make for less nice urls (seo people love to have nice image urls with their words in it) but we could do that. we could also allow random words to come after the format, something like /{id}/{format}/{random-seo-blah}.png. if we need a major version bump for that, go for it. if you want to remove phpcr-odm support, so be it - i am glad to give some input if we can keep it in, but can't commit to do a lot of work to maintain the support.

**Comment by greg0ire** [comment]
> i would not be aware of a decision to deprecate the phpcr-odm integration here. what we atm don't maintain anymore is the https://github.com/symfony-cmf/media-bundle which builds on top of the sonata media bundle.

Damn. Back to square one then.

> in phpcr-odm, document ids indeed are paths with forward slashes. the alternative would be to use the uuid of the documents which is a uuid v4. this will make for less nice urls (seo people love to have nice image urls with their words in it) but we could do that. we could also allow random words to come after the format, something like /{id}/{format}/{random-seo-blah}.png. if we need a major version bump for that, go for it. if you want to remove phpcr-odm support, so be it - i am glad to give some input if we can keep it in, but can't commit to do a lot of work to maintain the support.

That would mean using `findByUuid` just for this manager, right? What about my solution, using `/{phpcr-seo-friendly-id-with-forward-slashes}?format={format}` ? It won't require complicated logic for [this line](https://github.com/sonata-project/SonataMediaBundle/blob/3.x/Controller/MediaController.php#L41)

**Comment by dbu** [comment]
> What about my solution, using
/{phpcr-seo-friendly-id-with-forward-slashes}?format={format}

moving the format into the query string could work i guess, though its a
bit ugly. saving the file is less nice (what extension for `image` with
no extension in the path?) and i just see somebody ending up with
/my/path/image.png?format=jpg

> That would mean using |findByUuid| just for this manager, right?

no, phpcr-odm detects uuid in find() and will return the document
regardless of whether it was specified by path or by uuid.

i think either has its merits and drawbacks. i'd give my vote to format
in the query string, but take this as a "weak" vote as i am not very
convinced its the best solution

**Comment by greg0ire** [comment]
> no, phpcr-odm detects uuid in find() and will return the document regardless of whether it was specified by path or by uuid.

Oh, good to know. I guess your solution should be quite fine, in that case.

> and i just see somebody ending up with /my/path/image.png?format=jpg

`format` is supposed to map to a size ('big', 'small', 'reference'), not to a file extension.

I think both solution are quite fine, and that we should let @NeuralClone decide what he thinks is best, since he probably has more context in memory.

**Comment by NeuralClone** [comment]
First of all, thanks for giving some input on this, @dbu. Some good alternative suggestions. On some level, it's unfortunate that PHPCR-ODM is still being supported because it complicates the fix.

This ended up being a bit longer than I intended it to be but I wanted give the full rationale for my suggested fix. Please bear with me. I am going somewhere with this. 😄 

I think I may have come up with a solution that could work. It's a little different than what's been suggested so far and isn't perfect. Basically, why not flip the order of the parameters in the route configuration? I haven't fully tested this yet but I think it'd take the least amount of work to implement and is fairly intuitive since it works very similar to what we already have.

```xml
<route id="sonata_media_view" path="/view/{format}/{id}">
    <default key="_controller">SonataMediaBundle:Media:view</default>
    <default key="format">reference</default>
    <requirement key="format">[^/]+</requirement>
    <requirement key="id">.+</requirement>
</route>

<route id="sonata_media_download" path="/download/{format}/{id}">
    <default key="_controller">SonataMediaBundle:Media:download</default>
    <default key="format">reference</default>
    <requirement key="format">[^/]+</requirement>
    <requirement key="id">.+</requirement>
</route>

<route id="sonata_media_gallery_view" path="/view/{id}">
    <default key="_controller">SonataMediaBundle:Gallery:view</default>
    <requirement key="id">.+</requirement>
</route>
```

Right now the problem is that `format` isn't being filled in all situations when it should be. It's impossible to fill it when PHPCR-ODM is in use. Since PHPCR-ODM is still being maintained, it isn't acceptable to not allow slashes in the `id` as they're critical to how it handles `id` values. I also don't think it makes sense to remove support for PHPCR-ODM just to fix this bug. Doing that would eliminate some of the flexibility of the bundle.

As part of my proposed solution, I'd also suggest tightening down the requirements slightly on both `format` and `id` to prevent empty parameters from appearing in `id` and `format` and slashes from appearing in `format`. Only preventing slashes in `format` is absolutely required for this to work, though.

Some pros and cons for my proposed solution.

**Pros**
1) The Routing component maps route parameters to a controller's parameters by name, not by the order they appear. So the signature of the controllers could remain as they are now, which would promote BC.

2) We could keep PHPCR-ODM support.

3) `format` would always be filled.

4) We could avoid using query strings (i.e., keep consistent "pretty" URLs).

5) A very small change would be required in the way routes are generated and handled.

6) No need to inject the RequestStack service or other services into controllers or helper services.

7) Consistency between the way the `sonata_media_view` and `sonata_media_download` routes work. (Currently the view route doesn't properly support PHPCR-ODM `id` values that include slashes.)

**Cons**
1) The `format` parameter would become mandatory. Even if it's automatically filled in most situations by the bundle, it would have to appear in every path that's generated by the Routing component. For example: `/view/reference/my/phpcr/odm/id` instead of the current `/view/my/phpcr/odm/id`.

2) The default format, `reference`, isn't very intuitive as part of URLs that users would be interacting with.

I do have an idea for addressing the second con, though. We could create an alias for "reference" that would only be used in generated paths. I'd suggest "original" to represent the original format of the media. From a user's point of view, "original" would work like "reference". Internally, a few checks would need to be made but other than that the bundle would operate the same way it does now.

In other words, to users `/view/reference/my/phpcr/odm/id` would become `/view/original/my/phpcr/odm/id`. "original" would become an alias for "reference". Both would work through paths but only the version with "original" would appear to users. Functionally these two values would operate the same way.

Using "reference" in place of "original" shouldn't present any sort of security concern either since "reference" already works the way I'm suggesting "original" would. The one problem I see with making "original" a reserved `format` name is it could result in a BC-break if people have configured a format with that name. But that seems like a minor concern given the benefits these changes would provide.

**Comment by greg0ire** [comment]
I thought of it, but didn't propose it because I see more cons:

- Making the `format` parameter mandatory is a BC-break for people using the route directly.
- It's more natural IMO to say that the file as several formats than to say that there are many files in the `big` format, which means that putting the `format` first isn't very natural either.

Additionally, I don't know why you guys think urls with a query string are less pretty. I'd says that paths are great for expressing a clear hierarchy, like `/folder/subfolder/file`, but when you have parameters that are more like tags you put on a given resource, using `file?size=big&type=grayscale&ratio=16-9' feels a bit more logical.

With that approach : 

- format doesn't need to be always provided
- you have no url BC break
- tell me if I'm wrong, but I think `path()` uses the query string for anything that does not appear in the url pattern, so no change in url generation either

So the only con IMO would be the need to add a `$request` argument to the controller (no need to access the request stack, right? This would be a BC-break, but a very small one : it only breaks controllers extending the media controller.

**Comment by NeuralClone** [comment]
Oops, I forgot to include that first con that you mentioned. I agree that that's a problem and would definitely be a BC-break.

I think you raise several good points too. I can't really disagree with any of them. At the same time, I still think switching the order of the parameters is still a fairly intuitive way to approach things but I can also see why you disagree with that. There's a lot to be said for using a clear hierarchy.

You're also correct about `path()`/`Router::generate()`. Any parameter not included in the route configuration is appended to the URL as part of the query string. As for what's considered "pretty," I think that comes down to personal preference. I can see a valid case made both for and against using them.

Unfortunately, injecting `Request` directly is deprecated. So if we want to be up to date, we have to inject the `RequestStack` service and then retrieve the current request. That isn't a huge deal by any means. It just adds a little more work.

Ultimately, I'm fine with whatever approach people think is best. :)

**Comment by greg0ire** [comment]
> Ultimately, I'm fine with whatever approach people think is best. :)

Ok, great! @sonata-project/contributors , can you please help us take a decision? What do you think would be best?

**Comment by dbu** [comment]
i don't think expecting the request in the controller action is deprecated. see also https://symfony.com/doc/current/controller.html#the-request-object-as-a-controller-argument . what is deprecated in symfony 2 and no longer possible in 3 is using the `request` service, because that created a ton of problems with temporality and subrequests.

my gut feeling would have been to put the format before the id and use original by default. but if people are expected to call the path/generate function directly, that is indeed a BC break. i actually can't think of a good argument why the format should not be in the query string. so my vote goes for query string. 

we could keep a differently named BC route definition with the format at the end of the path, to not mess up on deployment with cached pages and such. and then i think this can be done completely BC except for when people extended the controller. we could even add a new controller method instead of changing the existing, that calls the old method with the format extracted from the request, to avoid a BC break alltogether.

**Comment by NeuralClone** [comment]
> i don't think expecting the request in the controller action is deprecated.

Hmm, it seems you're right. I must have been thinking of the service.

**Comment by NeuralClone** [comment]
Anyone else have any thoughts on the suggested fixes?
